psycopg/psycopg/connection.py

```diff
@@ -95,27 +95,44 @@ def connect(
         timeout = timeout_from_conninfo(params)
         rv = None
         attempts = conninfo_attempts(params)
+        connection_errors: list[tuple[e._NoTracebackException, str]] = []
         for attempt in attempts:
             try:
                 conninfo = make_conninfo("", **attempt)
                 gen = cls._connect_gen(conninfo, timeout=timeout)
                 rv = waiting.wait_conn(gen, interval=_WAIT_INTERVAL)
             except e._NO_TRACEBACK as ex:
-                if len(attempts) > 1:
-                    logger.debug(
-                        "connection attempt failed: host: %r port: %r, hostaddr %r: %s",
-                        attempt.get("host"),
-                        attempt.get("port"),
-                        attempt.get("hostaddr"),
-                        str(ex),
-                    )
-                last_ex = ex
+                attempt_details = "host: {}, port: {}, hostaddr: {}".format(
+                    repr(attempt.get("host")),
+                    repr(attempt.get("port")),
+                    repr(attempt.get("hostaddr")),
+                )
+                connection_errors.append((ex, attempt_details))
             else:
                 break
 
         if not rv:
-            assert last_ex
-            raise last_ex.with_traceback(None)
+            num_errors = len(connection_errors)
+            assert num_errors > 0
+
+            last_exception, _ = connection_errors[-1]
+            if num_errors == 1:
+                raise last_exception.with_traceback(None)
+            else:
+                formatted_attempts = []
+                for error, attempt_details in connection_errors:
+                    formatted_attempts.append(f"- {attempt_details}: {str(error)}")
+                # Create a new exception with the same type as the last one, containing
+                # all attempt errors while preserving backward compatibility.
+                last_exception_type = type(last_exception)
+                message_lines = [
+                    f"{last_exception}",
+                    "Multiple connection attempts failed. All failures were:",
+                ]
+                message_lines.extend(formatted_attempts)
+                enhanced_message = "\n".join(message_lines)
+                enhanced_exception = last_exception_type(enhanced_message)
+                raise enhanced_exception.with_traceback(None) from last_exception
```

Comment:
Does it make sense to raise `from last_exception`? I don't think it adds much value. Maybe a simple exception is better.

Reply:
I think this is a hold-over from before you mentioned they were already flattened, so I was trying to preserve the origin of the exception. Since they're flat, I don't think this makes sense. Will remove.

Reply:
Addressed in [982dda4](https://github.com/psycopg/psycopg/commit/982dda4f5c99832d95e0e447ba6c16eb9122a24a)